product_internal_reference_generator/tests/test_product_internal_reference_generator.py

```diff
@@ -0,0 +1,70 @@
+            "product_internal_reference_generator.group_int_ref_template_always_visible"
+        )
+        cls.pt_seq = cls.env.ref("product_internal_reference_generator.demo_pcs_1")
+        cls.attr1 = attr_model.create({"name": "Wheels", "create_variant": "always"})
+        cls.attr2 = attr_model.create({"name": "Frame"})
+        cls.val1 = val_model.create({"name": "Big", "attribute_id": cls.attr1.id})
+        cls.val2 = val_model.create({"name": "Very Big", "attribute_id": cls.attr1.id})
+        cls.val3 = val_model.create({"name": "Carbon", "attribute_id": cls.attr2.id})
+        cls.val4 = val_model.create({"name": "Magnesium", "attribute_id": cls.attr2.id})
+        cls.pt_bicycle = pt_model.create({"name": "Bicycle"})
+        cls.pt_car = pt_model.create({"name": "Car"})
+        cls.pt_plane = pt_model.create({"name": "Plane"})
+
+    def test_all(self):
+        self.assertFalse(self.pt_bicycle.default_code)
+        self.pt_bicycle.int_ref_template_id = self.pt_seq
+        self.pt_bicycle.onchange_int_ref_template_id()
+        self.assertFalse(self.pt_bicycle.variants_prefix)
+        self.pt_bicycle.btn_generate_sequence()
+        df = str(self.pt_seq.sequence_id.number_next_actual - 1) + "000"
+        self.assertIn(df, self.pt_bicycle.product_variant_id.default_code)
+        self.pt_car.int_ref_template_id = self.pt_seq
+        self.pt_car.attribute_line_ids = [
+            (
+                0,
+                0,
+                {
+                    "attribute_id": self.attr1.id,
+                    "value_ids": [(6, 0, [self.val1.id, self.val2.id])],
+                },
+            ),
+        ]
+        self.pt_car.btn_generate_sequence()
+        df = str(self.pt_seq.sequence_id.number_next_actual - 1) + "001"
+        self.assertIn(df, self.pt_car.product_variant_ids[0].default_code)
+        df = str(self.pt_seq.sequence_id.number_next_actual - 1) + "002"
+        self.assertIn(df, self.pt_car.product_variant_ids[1].default_code)
+
+        self.pt_car.attribute_line_ids = [
+            (
+                0,
+                0,
+                {
+                    "attribute_id": self.attr2.id,
+                    "value_ids": [(6, 0, [self.val3.id, self.val4.id])],
+                },
+            ),
+        ]
+        df = str(self.pt_seq.sequence_id.number_next_actual - 1) + "003"
+        self.assertIn(df, self.pt_car.product_variant_ids[0].default_code)
+        df = str(self.pt_seq.sequence_id.number_next_actual - 1) + "004"
+        self.assertIn(df, self.pt_car.product_variant_ids[1].default_code)
+        df = str(self.pt_seq.sequence_id.number_next_actual - 1) + "005"
+        self.assertIn(df, self.pt_car.product_variant_ids[2].default_code)
+        df = str(self.pt_seq.sequence_id.number_next_actual - 1) + "006"
+        self.assertIn(df, self.pt_car.product_variant_ids[3].default_code)
+        self.assertFalse(self.pt_plane.get_variant_next_default_code())
```

Comment:
**suggestion(non-blocking)**: consider adding a test that uses the Form class (also so you don't have to call onchanges manually)